plugins/parquet-store/parquet_store.cpp

```diff
@@ -0,0 +1,564 @@
+#include <arrow/io/file.h>
+#include <arrow/io/memory.h>
+#include <arrow/ipc/dictionary.h>
+#include <arrow/ipc/reader.h>
+#include <arrow/status.h>
+#include <arrow/table.h>
+#include <arrow/type.h>
+#include <arrow/type_fwd.h>
+#include <arrow/util/key_value_metadata.h>
+#include <caf/actor_system_config.hpp>
+#include <caf/attach_stream_sink.hpp>
+#include <caf/expected.hpp>
+#include <caf/typed_event_based_actor.hpp>
+#include <parquet/arrow/reader.h>
+#include <parquet/arrow/writer.h>
+
+#include <memory>
+#include <utility>
+
+namespace vast::plugins::parquet_store {
+
+std::filesystem::path store_path_for_partition(const uuid& partition_id) {
+  auto store_filename = fmt::format("{}.parquet", partition_id);
+  return std::filesystem::path{"archive"} / store_filename;
+}
+
+std::optional<std::shared_ptr<arrow::Array>>
+fix_enum_array(const enumeration_type& et,
+               const std::shared_ptr<arrow::Array>& arr) {
+  switch (arr->type_id()) {
+    case arrow::Type::DICTIONARY: {
+      const auto dict = static_pointer_cast<arrow::DictionaryArray>(arr);
+      auto indices = arrow::compute::Cast(*dict->indices(), arrow::uint8());
+      return enumeration_type::array_type::make(
+               et.to_arrow_type(), std::static_pointer_cast<arrow::UInt8Array>(
+                                     indices.ValueOrDie()))
+        .MoveValueUnsafe();
+    }
+    case arrow::Type::STRING: {
+      auto values = static_pointer_cast<arrow::StringArray>(arr);
+      auto builder = enumeration_type::builder_type(et.to_arrow_type());
+      if (!builder.Reserve(values->length()).ok())
+        die("failed to reserve builder capacity for dict indices");
+      for (auto v : *values) {
+        if (v) {
+          if (!builder.Append(*et.resolve(v->to_string())).ok())
+            die("unable to append dict value");
+        } else {
+          if (!builder.AppendNull().ok())
+            die("unable to append null to dict indices");
+        }
+      }
+      return builder.Finish().ValueOrDie();
+    }
+    default: {
+      die(fmt::format("unhandled enum-parquet variation for array type '{}'",
+                      arr->type()->ToString()));
+      return arr;
+    }
+  }
+}
+
+/// Transform a chunked array by applying a mapping function `Mapper` over each
+/// chunk and constructs a new array from the transformed chunks.
+template <typename Mapper, class VastType>
+std::optional<std::shared_ptr<arrow::ChunkedArray>>
+map_chunked_array(const VastType& t,
+                  const std::shared_ptr<arrow::ChunkedArray>& arr, Mapper m) {
+  auto chunks = arrow::ArrayVector{};
+  chunks.reserve(arr->num_chunks());
+  for (const auto& chunk : arr->chunks()) {
+    if (std::optional<std::shared_ptr<arrow::Array>> c = m(t, chunk))
+      chunks.push_back(*c);
+    else
+      return arr;
+  }
+  auto result = arrow::ChunkedArray::Make(chunks);
+  return result.ValueOrDie();
+}
+
+std::optional<std::shared_ptr<arrow::Array>>
```

Comment:
I think it's okay. Do we really need the second level of nullability, though? Doesn't a shared_ptr already suffice as you could just return nullptr instead of nullopt?

Reply:
That's a great point, didn't think of that. I stripped out all those `std::optionals` and things just keep working, it's magical.